scripts/exp/create-overloads.ts

```diff
@@ -0,0 +1,321 @@
+  const sourceFile = program.getSourceFile(input)!;
+
+  const result = ts.transform<ts.SourceFile>(sourceFile, [
+    keepPublicFunctionsTransformer.bind(
+      undefined,
+      program,
+      replace,
+      moduleToEnhance
+    )
+  ]);
+
+  const transformedSourceFile = result.transformed[0];
+  const content = printer.printFile(transformedSourceFile);
+
+  // if append, append to the output file
+  if (append) {
+    if (!fs.existsSync(output)) {
+      throw Error(
+        `${output} doesn't exist. Please provide path to an existing file when using the -a option`
+      );
+    }
+    const stat = fs.statSync(output);
+    if (!stat.isFile()) {
+      throw Error(
+        `${output} is not a file. Please provide path to an existing file when using the -a option`
+      );
+    }
+
+    fs.appendFileSync(output, `\n${content}`);
+  } else {
+    fs.writeFileSync(output, content);
+  }
+}
+
+function keepPublicFunctionsTransformer(
+  program: ts.Program,
+  replace: ReplaceOption[],
+  moduleNameToEnhance: string,
+  context: ts.TransformationContext
+): ts.Transformer<ts.SourceFile> {
+  return (sourceFile: ts.SourceFile) => {
+    const typeChecker = program.getTypeChecker();
+    const overloads: ts.Statement[] = [];
+    function visit(node: ts.Node): ts.Node {
+      if (ts.isFunctionDeclaration(node)) {
+        // return early if the function doesn't have any parameter of the type we are looking for
+        if (
+          !node.parameters.find(param => {
+            if (param.type && ts.isTypeReferenceNode(param.type)) {
+              const typeName = param.type.typeName;
+              return replace.find(opt => typeName.getText() === opt.match);
+            }
+            return false;
+          })
+        ) {
+          return ts.createToken(ts.SyntaxKind.WhitespaceTrivia);
+        }
+
+        const newParameters = node.parameters.map(param => {
+          if (param.type && ts.isTypeReferenceNode(param.type)) {
+            for (const replaceOption of replace) {
+              if (
+                param.type.typeName.getText(sourceFile) === replaceOption.match
+              ) {
+                return ts.updateParameter(
+                  param,
+                  param.decorators,
+                  param.modifiers,
+                  param.dotDotDotToken,
+                  param.name,
+                  param.questionToken,
+                  ts.createTypeReferenceNode(
+                    replaceOption.replacement,
+                    param.type.typeArguments
+                  ),
+                  param.initializer
+                );
+              }
+            }
+          }
+
+          return param;
+        });
+
+        // remove comments
+        ts.setTextRange(node, {
+          pos: node.getStart(sourceFile),
+          end: node.getEnd()
+        });
+
+        overloads.push(
+          ts.updateFunctionDeclaration(
+            node,
+            node.decorators,
+            [],
```

Comment:
Deliberately removing any modifiers? What are modifiers? "public|private|protected"?

Reply:
Yes, to remove `export` and `declare` which are not needed in the `declare module` block